File: lib/js/src/unicode.js
// The character data of grammars/unicode.txt: the class of a character and
// the simple lowercase mapping, the same in every gencmu library.

export class UnicodeTable {
  /** @param {string} text the contents of unicode.txt */
  constructor(text) {
    /** @type {string | null} */
    this.version = null;
    /** @type {[number, number][]} */
    this.marks = [];
    /** @type {[number, number][]} */
    this.alphas = [];
    /** @type {Map<number, number>} */
    this.lower = new Map();
    for (const line of text.split("\n")) {
      const fields = line.trim().split(/\s+/);
      if (fields[0] === "unicode") this.version = fields[1];
      else if (fields[0] === "mark") this.marks.push([parseInt(fields[1], 16), parseInt(fields[2], 16)]);
      else if (fields[0] === "alpha") this.alphas.push([parseInt(fields[1], 16), parseInt(fields[2], 16)]);
      else if (fields[0] === "lower") this.lower.set(parseInt(fields[1], 16), parseInt(fields[2], 16));
    }
  }

  /**
   * The class tag of a code point (engine §1).
   * @param {number} code
   * @returns {"space" | "digit" | "mark" | "alpha" | "other"}
   */
  classOf(code) {
    if ((code >= 0x09 && code <= 0x0d) || code === 0x20 || code === 0x85 || code === 0xa0 || code === 0x1680 ||
        (code >= 0x2000 && code <= 0x200a) || code === 0x2028 || code === 0x2029 || code === 0x202f ||
        code === 0x205f || code === 0x3000) {
      return "space";
    }
    if (code >= 0x30 && code <= 0x39) return "digit";
    if (inRanges(this.marks, code)) return "mark";
    if (inRanges(this.alphas, code)) return "alpha";
    return "other";
  }

  /**
   * @param {string} text
   * @returns {string}
   */
  lowercase(text) {
    let result = "";
    for (const character of text) {
      const mapped = this.lower.get(/** @type {number} */ (character.codePointAt(0)));
      result += mapped === undefined ? character : String.fromCodePoint(mapped);
    }
    return result;
  }
}

/**
 * @param {[number, number][]} ranges
 * @param {number} code
 */
function inRanges(ranges, code) {
  let low = 0;
  let high = ranges.length - 1;
  while (low <= high) {
    const middle = (low + high) >> 1;
    const [start, end] = ranges[middle];
    if (code < start) high = middle - 1;
    else if (code > end) low = middle + 1;
    else return true;
  }
  return false;
}

File: lib/rust/src/unicode.rs
//! The character table of `grammars/unicode.txt` (engine §1): the `mark`
//! and `alpha` ranges that give a character its class, and the simple
//! lowercase mappings `lowercase()` uses. Every library reads this table
//! rather than its platform's Unicode data, so that all agree.

use std::collections::HashMap;

#[derive(Debug)]
pub(crate) struct Unicode {
    marks: Vec<(u32, u32)>,
    alphas: Vec<(u32, u32)>,
    lower: HashMap<u32, u32>,
}

fn hex(field: Option<&str>, line: usize) -> Result<u32, String> {
    let field = field.ok_or_else(|| format!("unicode.txt line {line}: a missing field"))?;
    u32::from_str_radix(field, 16).map_err(|_| format!("unicode.txt line {line}: a bad code point {field:?}"))
}

impl Unicode {
    pub(crate) fn parse(text: &str) -> Result<Unicode, String> {
        let mut table = Unicode { marks: Vec::new(), alphas: Vec::new(), lower: HashMap::new() };
        for (index, line) in text.lines().enumerate() {
            let number = index + 1;
            let mut fields = line.split_whitespace();
            match fields.next() {
                None | Some("unicode") => {}
                Some("mark") => table.marks.push((hex(fields.next(), number)?, hex(fields.next(), number)?)),
                Some("alpha") => table.alphas.push((hex(fields.next(), number)?, hex(fields.next(), number)?)),
                Some("lower") => {
                    let from = hex(fields.next(), number)?;
                    let to = hex(fields.next(), number)?;
                    table.lower.insert(from, to);
                }
                Some(other) => return Err(format!("unicode.txt line {number}: an unknown entry {other:?}")),
            }
        }
        table.marks.sort_unstable();
        table.alphas.sort_unstable();
        Ok(table)
    }

    fn within(ranges: &[(u32, u32)], code: u32) -> bool {
        let index = ranges.partition_point(|&(first, _)| first <= code);
        index > 0 && ranges[index - 1].1 >= code
    }

    /// The weak class tag of a character (engine §1).
    pub(crate) fn class(&self, c: char) -> &'static str {
        let code = c as u32;
        let space = matches!(
            code,
            0x09..=0x0D | 0x20 | 0x85 | 0xA0 | 0x1680 | 0x2000..=0x200A | 0x2028 | 0x2029 | 0x202F | 0x205F | 0x3000
        );
        if space {
            "space"
        } else if (0x30..=0x39).contains(&code) {
            "digit"
        } else if Self::within(&self.marks, code) {
            "mark"
        } else if Self::within(&self.alphas, code) {
            "alpha"
        } else {
            "other"
        }
    }

    /// The simple lowercase mapping of a string, code point by code point.
    pub(crate) fn lowercase(&self, text: &str) -> String {
        text.chars().map(|c| self.lower.get(&(c as u32)).and_then(|&to| char::from_u32(to)).unwrap_or(c)).collect()
    }
}
